File: Test/AdventureWorksFunctionalModel/Mapping/CurrencyRateMap.cs
using System.Data.Entity.ModelConfiguration;

using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;

namespace AW.Mapping {
    public class CurrencyRateMap : EntityTypeConfiguration<CurrencyRate> {
        public CurrencyRateMap() {
            // Primary Key
            HasKey(t => t.CurrencyRateID);

            // Properties
            Property(t => t.FromCurrencyCode)
                .IsRequired()
                .IsFixedLength()
                .HasMaxLength(3);

            Property(t => t.ToCurrencyCode)
                .IsRequired()
                .IsFixedLength()
                .HasMaxLength(3);

            // Table & Column Mappings
            ToTable("CurrencyRate", "Sales");
            Property(t => t.CurrencyRateID).HasColumnName("CurrencyRateID");
            Property(t => t.CurrencyRateDate).HasColumnName("CurrencyRateDate");
            Property(t => t.FromCurrencyCode).HasColumnName("FromCurrencyCode");
            Property(t => t.ToCurrencyCode).HasColumnName("ToCurrencyCode");
            Property(t => t.AverageRate).HasColumnName("AverageRate");
            Property(t => t.EndOfDayRate).HasColumnName("EndOfDayRate");
            Property(t => t.ModifiedDate).HasColumnName("ModifiedDate"); //.IsConcurrencyToken();

            // Relationships
            HasRequired(t => t.Currency).WithMany().HasForeignKey(t => t.FromCurrencyCode);
            HasRequired(t => t.Currency1).WithMany().HasForeignKey(t => t.ToCurrencyCode);
        }
    }

    public static partial class Mapper {
        public static void Map(this EntityTypeBuilder<CurrencyRate> builder) {
            builder.HasKey(t => t.CurrencyRateID);

            // Properties
            builder.Property(t => t.FromCurrencyCode)
                   .IsRequired()
                   .IsFixedLength()
                   .HasMaxLength(3);

            builder.Property(t => t.ToCurrencyCode)
                   .IsRequired()
                   .IsFixedLength()
                   .HasMaxLength(3);

            // Table & Column Mappings
            builder.ToTable("CurrencyRate", "Sales");
            builder.Property(t => t.CurrencyRateID).HasColumnName("CurrencyRateID");
            builder.Property(t => t.CurrencyRateDate).HasColumnName("CurrencyRateDate");
            builder.Property(t => t.FromCurrencyCode).HasColumnName("FromCurrencyCode");
            builder.Property(t => t.ToCurrencyCode).HasColumnName("ToCurrencyCode");
            builder.Property(t => t.AverageRate).HasColumnName("AverageRate");
            builder.Property(t => t.EndOfDayRate).HasColumnName("EndOfDayRate");
            builder.Property(t => t.ModifiedDate).HasColumnName("ModifiedDate"); //.IsConcurrencyToken();

            // Relationships
            builder.HasOne(t => t.Currency).WithMany().HasForeignKey(t => t.FromCurrencyCode);
            builder.HasOne(t => t.Currency1).WithMany().HasForeignKey(t => t.ToCurrencyCode);
        }
    }
}

File: Test/AdventureWorksLegacyModelVB/Mapping/CurrencyRateMap.vb
Imports System.Data.Entity.ModelConfiguration

Imports Microsoft.EntityFrameworkCore
Imports Microsoft.EntityFrameworkCore.Metadata.Builders

Namespace AW.Mapping
	Public Class CurrencyRateMap
		Inherits EntityTypeConfiguration(Of CurrencyRate)

		Public Sub New()
			' Primary Key
			HasKey(Function(t) t.CurrencyRateID)

			' Properties
			[Property](Function(t) t.FromCurrencyCode).IsRequired().IsFixedLength().HasMaxLength(3)

			[Property](Function(t) t.ToCurrencyCode).IsRequired().IsFixedLength().HasMaxLength(3)

			' Table & Column Mappings
			ToTable("CurrencyRate", "Sales")
			[Property](Function(t) t.CurrencyRateID).HasColumnName("CurrencyRateID")
			[Property](Function(t) t.CurrencyRateDate).HasColumnName("CurrencyRateDate")
			[Property](Function(t) t.FromCurrencyCode).HasColumnName("FromCurrencyCode")
			[Property](Function(t) t.ToCurrencyCode).HasColumnName("ToCurrencyCode")
			[Property](Function(t) t.AverageRate).HasColumnName("AverageRate")
			[Property](Function(t) t.EndOfDayRate).HasColumnName("EndOfDayRate")
			[Property](Function(t) t.ModifiedDate).HasColumnName("ModifiedDate") '.IsConcurrencyToken();

			' Relationships
			HasRequired(Function(t) t.Currency).WithMany().HasForeignKey(Function(t) t.FromCurrencyCode)
			HasRequired(Function(t) t.Currency1).WithMany().HasForeignKey(Function(t) t.ToCurrencyCode)
		End Sub
	End Class

	Public Module Mapper
		<System.Runtime.CompilerServices.Extension> _
		Public Sub Map(ByVal builder As EntityTypeBuilder(Of CurrencyRate))
			builder.HasKey(Function(t) t.CurrencyRateID)

			' Properties
			builder.Property(Function(t) t.FromCurrencyCode).IsRequired().IsFixedLength().HasMaxLength(3)

			builder.Property(Function(t) t.ToCurrencyCode).IsRequired().IsFixedLength().HasMaxLength(3)

			' Table & Column Mappings
			builder.ToTable("CurrencyRate", "Sales")
			builder.Property(Function(t) t.CurrencyRateID).HasColumnName("CurrencyRateID")
			builder.Property(Function(t) t.CurrencyRateDate).HasColumnName("CurrencyRateDate")
			builder.Property(Function(t) t.FromCurrencyCode).HasColumnName("FromCurrencyCode")
			builder.Property(Function(t) t.ToCurrencyCode).HasColumnName("ToCurrencyCode")
			builder.Property(Function(t) t.AverageRate).HasColumnName("AverageRate")
			builder.Property(Function(t) t.EndOfDayRate).HasColumnName("EndOfDayRate")
			builder.Property(Function(t) t.ModifiedDate).HasColumnName("ModifiedDate") '.IsConcurrencyToken();

			' Relationships
			builder.HasOne(Function(t) t.Currency).WithMany().HasForeignKey(Function(t) t.FromCurrencyCode)
			builder.HasOne(Function(t) t.Currency1).WithMany().HasForeignKey(Function(t) t.ToCurrencyCode)
		End Sub
	End Module
End Namespace